source-sqlserver/replication.go

```diff
@@ -0,0 +1,488 @@
+	}
+
+	for rows.Next() {
+		if err := rows.Scan(vptrs...); err != nil {
+			return fmt.Errorf("error scanning result row: %w", err)
+		}
+		var fields = make(map[string]interface{})
+		for idx, name := range cnames {
+			fields[name] = vals[idx]
+		}
+
+		log.WithFields(log.Fields{"stream": info.StreamID, "data": fields}).Trace("got change")
+
+		// Extract the event's LSN and remove it from the record fields, since
+		// it will be included in the source metadata already.
+		var changeLSN, ok = fields["__$start_lsn"].([]byte)
+		if !ok {
+			changeLSN = nil
+		}
+		if bytes.Compare(changeLSN, fromLSN) <= 0 {
+			log.WithFields(log.Fields{
+				"stream":    info.StreamID,
+				"changeLSN": changeLSN,
+				"fromLSN":   fromLSN,
+			}).Trace("filtered change <= fromLSN")
+			continue
+		}
+		delete(fields, "__$start_lsn")
+
+		// Extract the change type (Insert/Update/Delete) and remove it from the record data.
+		var operation sqlcapture.ChangeOp
+		opval, ok := fields["__$operation"]
+		if !ok {
+			return fmt.Errorf("internal error: cdc operation unspecified ('__$operation' unset)")
+		}
+		opcode, ok := opval.(int64)
+		if !ok {
+			return fmt.Errorf("internal error: cdc operation code %T(%#v) isn't an int64", opval, opval)
+		}
+		switch opcode {
+		case 1:
+			operation = sqlcapture.DeleteOp
+		case 2:
+			operation = sqlcapture.InsertOp
+		case 4:
+			operation = sqlcapture.UpdateOp
+		default:
+			return fmt.Errorf("invalid change operation: %d", opcode)
+		}
+		delete(fields, "__$operation")
+
+		// Move the '__$seqval' and '__$update_mask' columns into metadata as well.
+		var seqval = fields["__$seqval"]
+		var updateMask = fields["__$update_mask"]
+		delete(fields, "__$seqval")
+		delete(fields, "__$update_mask")
+
+		var rowKey, err = sqlcapture.EncodeRowKey(info.KeyColumns, fields, info.ColumnTypes, encodeKeyFDB)
+		if err != nil {
+			return fmt.Errorf("error encoding stream %q row key: %w", info.StreamID, err)
+		}
+		if err := rs.db.translateRecordFields(info.ColumnTypes, fields); err != nil {
+			return fmt.Errorf("error translating stream %q change event: %w", info.StreamID, err)
+		}
+
+		// For inserts and updates the change event records the state after the
+		// change, and for deletions the change event (partially) records the state
+		// before the change.
+		var before, after map[string]any
+		if operation == sqlcapture.InsertOp || operation == sqlcapture.UpdateOp {
+			after = fields
+		} else {
+			before = fields
+		}
+
+		rs.events <- &sqlcapture.ChangeEvent{
+			Operation: operation,
+			RowKey:    rowKey,
+			Source: &sqlserverSourceInfo{
+				SourceCommon: sqlcapture.SourceCommon{
+					Schema: info.SchemaName,
+					Table:  info.TableName,
+				},
+				LSN:        changeLSN,
+				SeqVal:     seqval,
+				UpdateMask: updateMask,
+			},
+			Before: before,
+			After:  after,
+		}
+	}
+
+	return nil
+}
+
+func cdcGetMinLSN(ctx context.Context, conn *sql.DB, instanceName string) ([]byte, error) {
```

Comment:
It isn't used currently, but at some point in the not too distant future we'll probably want an explicit "is our saved cursor LSN still available in the change tables?" sanity check at startup so I didn't delete the helper function when I modified things to start streaming from the *newest* LSN the first time the connector starts up.

Which, yeah, I guess we could delete it and rewrite it / retrieve it from the commit history later on, but I figured I'd just leave it there because I'm lazy.